src/dams_mcda/WSM.R

```diff
@@ -40,190 +45,386 @@ available_dams <- seq(1:8)
+	NormalizedMatrix <- array(data=NA, dim = c(8,14,995))
+	# array of rows that use minimization (cost or damage-related)
+	min_crit_columns <- c(4, 5, 6) 
+	
+	# make normalized values of each value in matrix 
+	for (k in 1:matrix_cols){
+	  for (n in 1:matrix_rows){
+	    for (p in 1:matrix_levs){
+	      x <- DamsDataMatrix[n,k,p]
+	      crit_min_x <- CritMinVector[k]
+	      crit_max_x <- CritMaxVector[k]
+	      
+	      NormalizedMatrix[n,k,p] <- tryCatch({ 
+	        
+	        if (k %in% min_crit_columns){
+	          # alternative method
+	          # maximize normalization
+	          ((1-(x-crit_max_x)) / (crit_max_x - crit_min_x))
+	        }else{
+	          # for debugging by cell WSM uncomment next line
+	          # message('cell n, k, x, crit, result', n, ', ', k, ', ', x, ', ', ', ', (((x - crit_min_x) / (crit_max_x - crit_min_x))) )
+	          
+	          # default method
+	          # minimize normilization
+	          ((x - crit_min_x) / (crit_max_x - crit_min_x))
+	        }
+	      }, error=function(e){
+	        (NA)
+	      })
+	    }
+	  }
 	}
+	
+	NormalizedMatrix[is.nan(NormalizedMatrix)] <-0
+	
+	message('Data column ', DamsDataMatrix[k])
+	message('Normalized column ', NormalizedMatrix[k])
+	
+	# debug
+	#message('NormalizedMatrix ', NormalizedMatrix)
+	
+	
+	#----------------------------------------
+	# WeightedScoreMatrix
+	
+	#----------------------------------------
+	
+	WeightedScoreMatrix <- (NormalizedMatrix*PrefMatrix)
+	
+	WeightedScoreMatrix <- round(WeightedScoreMatrix,3) 
+	
+	#----------------------------------------
+	# Score Sum
+	
+	# Weighted sum procedure:
+	#  multiply all normalized scores by prefrence weights
+	#  sum normalized, weighted criteria scores for each dam
+	#  sum this across all dams for each scenario
+	#-----------------------------------------
+	
+	#declare a weighted sum variable
+	scoresum_total <- rep(0,dim(NormalizedMatrix)[3])
+	
+	#multiply crit scores by user preferences
+	for (i in 1:dim(NormalizedMatrix)[3]){
+	  WSMMatrix <- NormalizedMatrix[,,i] * PrefMatrix[,,i]
+	  scoresum_total[i] <- sum(WSMMatrix) #this sums everything in each scenario after they are preferenced. Should be fine as order doesn't matter at this point.
+	}
+	
+	#-----------------------------------------
+	# Rank:
+	#  may need to reshape the array produced by the weighted sum procedure
+	#  sort the array by descending order, highest score comes first, and record the indices of the top ranked scenario
+	# Retrieve table, map of highest ranked scenario:
+	#  use server url or whatever, searching for map name with the matching index number
+	#  take the map image, table, and stick them in the webpage
+	#----------------------------------------
 
-	scoresum <- unlist(scoresum)
-
+	#order scenarios by rank: largest score first
+	idxRank <- order(scoresum_total,decreasing=TRUE)
+	
+	
+	#use scenario idxRank[1] to find corresponding map name
+	fname <- sprintf('maps/Penobscot_MO_14_%d.png',idxRank[1])

```

Comment:
should this commit include these files or is this placeholder for now

Reply:
yes they should be included. I'm adding them today, if they haven't yet been added.